src/compiler/crystal/semantic/ast.cr

```diff
@@ -13,6 +13,10 @@ module Crystal
       ::raise exception_type.for_node(self, message, inner)
     end
 
+    def unreachable!(message = "unreachable", inner = nil, exception_type = Crystal::TypeException)
+      ::raise exception_type.for_node(self, "BUG: #{message}", inner)
```

Comment:
It is difficult to raise `UnreachableError` with keeping node information. Perhaps we will feel confused in future in debug if disappear node trace.